### ISSUE TITLE
Provide multicast/subscription capabilities

### ISSUE THREAD

**Issue by eloylp** [eloylp/goomerang#3]
### Problem

Currently there are no explicit multicast/subscription capabilities in Goomerang. But they could be achieved in the following way:

```go
func main(){
    // Server instantiation ...
    // User creates a map for holding multicast groups
    groups := sync.Map{}
    s.Handle(&protos.SubscriptionV1{}, message.HandlerFunc(func(s message.Sender, msg *message.Message) {
               subs := msg.Payload.(*protos.SubscriptionV1)

		val, ok := groups.Load(subs.SubscribeTo)
		if !ok {
			var senders []message.Sender
			senders = append(senders, s)
                        // User saves the sender, into the new multicast group.
			groups.Store(subs.SubscribeTo, senders)
			return
		}
		senders := val.([]message.Sender)
		senders = append(senders, s)
		groups.Store(subs.SubscribeTo, senders)

}
```

In the above code example, the user prepares a especial handler listening for the `protos.SubscriptionV1` message,  which would carry the topic to subscribe to. Then the handler adds the current `message.Sender` to the previously created map. Each `message.Sender`, has the pointer of the underlying connection from which the subscription message arrived from.

Then, at the server, if the user wants to sends a multicast message, the code could be:

```go
func main() {
        // Recovering the group of connections ...
        senders, ok := groups.Load("group1")
	if !ok {
		panic("cannot found subscription group")
	}
	msgS := message.New().SetPayload(&protos.MessageV1{Message: "a message for group1"})
	ss := senders.([]message.Sender)
    // Sending through all the senders.
	for i := range ss {
		_, _ := ss[i].Send(msgS)
	}
}
```

That could work !! Now, what happens if the user wants to unsubscribe for a group for a specific connection ? Yes, the user could create another handler for unsubscribing a specific sender (using the pointer address). 

However, the above trick have some problems:

* Not enough intuitive for users.
* The user is on the wild. Needs to create all the logic externally to the library.
* The library is not providing other utilities around it, like metrics.

### Goals

The server and client should provide multi-cast and subscription capabilities in the public API. That will provide users ability to create interesting subscription models, without needing to build too much external logic.

Currently theres a Broadcast feature. Take it into account as some parts could be probably re-used.

Documentation should be written.

To not be invasive with the current design. The feature should be provided as a tool. Thus optional for the user.

**Comment by eloylp** [comment]
All right ! Lets start with this one. Initially looks like we would need to extend the goomerang internal protocol, by adding the following messages:

* `SubscribeV1`
* `UnsubscribeV1`
* `PublishV1`

The above messages would be the transport of the commands needed by the server. From the client public API perspective, having two methods `Subscribe(topic string)` and `Unsubscribe(topic string)` which would send the commented messages to the server, looks like the natural approach. 

Initially, all internal messages would be send asynchronously. If a client its already subscribed to a topic, a second call to `Subscribe(topic string)` would have no effect. The same would happen for `Unsubscribe(topic string)`.

The client and the server will implement a `Publish(topic string, msg *message.Message)` , which can be used to publish messages in a topic , reaching all the subscribed clients. In case of being called from the client API, the publish command will be sent to the server (using `PublishV1` internal message), which would do the multicast operation for the subscribed clients. In case of being called from the server API, the server will directly do the multicast operation, as it will have direct access to the connection map/grouping.

Talking about the server side, three built-in handlers would be needed for handling the commented messages `SubscribeV1` and `UnsubscribeV1`  and also the `PublishV1`. This handlers need to be available from the start.

This could a first happy path scenario. A first proposal. Lets see how it goes !

**Comment by eloylp** [comment]
Good ! looks like we can share the first protobuf representation of the internal messages:

```protobuf
message SubscribeCommand {
  string topic = 1;
}

message UnSubscribeCommand {
  string topic = 1;
}

message PublishCommand {
  string topic = 1;
  string kind = 2;
  bytes message = 3;
}
```
In case of the `PublishCommand` ,  we are embedding the user message in the `message` field. The `kind` field is necessary, it represent the protobuf FQDN, so the server can decode the user message to the concrete type. 

The headers of the `message` payload, will be copied to the `PublishCommand` headers, with the string `message-` as key prefix. When received by the server, headers would need to be extracted and added again to the original user message. In this way we can preserve the original headers.

A first test with [commented interfaces](https://github.com/eloylp/goomerang/issues/3#issuecomment-1186231890) is ready. Lets try to implement a basic POC !

**Comment by eloylp** [comment]
More progress on this :tada: . Looks like we have a quite good, harmonised implementation. [Following up the previous comments](https://github.com/eloylp/goomerang/issues/3#issuecomment-1186303519), we finally created the internal protocol buffers commands, so clients can perform `subscribe`, `publish` and `unsubscribe` operations. Such internal commands are not visible to the users of the library. Users of the library now have the following interfaces in the client public api:

```go

func (c *Client) Subscribe(topic string) (err error) {
	//...
}

func (c *Client) Publish(topic string, msg *message.Message) (payloadSize int, err error) {
	//...
}

func (c *Client) Unsubscribe(topic string) (err error) {
	//....
}
```
One decision that was made: Topics can transport multiple kind of messages. Although probably the normal use cases would not publish more than one kind of messages per topic. The reason for this is to not limit the user of the library. We leave that decision on their hands.

Before a client subscribes to a topic, all the messages need to have a defined handler or they will be ignored by the processing loop. This was made clear in the documentation.

The server side also has a `Publish` interface for server side publications. That would be helpful in case the server is connected to another source of data and wants to publish messages in the respective topics, which would be received by subscribed clients.

Okay, we talked about the public API surface. Lets now talk about the internal pub/sub engine. The initial idea is simple. We are going to hold a map of maps. The first level map will have the topic as key, and another map as value, which will use the connection slot as key and the respective sender as value:

```go
type pubSubEngine struct {
	csMap map[string]map[*conn.Slot]message.Sender
	L     *sync.RWMutex
}
```
The idea is to reduce the time spent in, as an example, removing a connection from the engine, as we will use the pointer address as a key for the map, probably reducing the domain in which the underlying hash function of the map needs to work. The problem could come when we try to iterate over the second level map for publishing messages, as it can be not as performant as we can imagine. Anyway, all of this are just dangerous assumptions. Another question would be how much time takes to write in a socket per each sender in the topic. Multiple improvements could be probably done in this area in a future.


Heres a PR holding this feature: https://github.com/eloylp/goomerang/pull/6
